File: java-algo/src/main/java/com/leetcode/Problem630.java
package com.leetcode;

import java.util.*;

/**
 * 630. Course Schedule III
 * https://leetcode.com/problems/course-schedule-iii/
 */
public class Problem630 {
    public int scheduleCourse(int[][] courses) {
        Arrays.sort(courses, Comparator.comparingInt(o -> o[1]));
        PriorityQueue<Integer> queue = new PriorityQueue<>((a, b) -> b - a);
        int days = 0;
        for (int[] course : courses) {
            days += course[0];
            queue.add(course[0]);
            if (days > course[1]) {
                days -= queue.poll();
            }
        }
        return queue.size();
    }
}

File: python-algo/leetcode/problem_630.py
"""630. Course Schedule III
https://leetcode.com/problems/course-schedule-iii/
"""
from typing import List
import heapq


class Solution:
    def schedule_course(self, courses: List[List[int]]) -> int:
        courses.sort(key=lambda x: x[1])
        days = 0
        learned = []
        heapq.heapify(learned)
        for course in courses:
            days += course[0]
            heapq.heappush(learned, -course[0])
            if days > course[1]:
                days += heapq.heappop(learned)
        return len(learned)
